src/rdzstd.c

```diff
@@ -0,0 +1,197 @@
+ *    and/or other materials provided with the distribution. 
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "rdkafka_int.h"
+#include "rdzstd.h"
+
+#include <zstd.h>
+#include <zstd_errors.h>
+
+rd_kafka_resp_err_t
+rd_kafka_zstd_decompress (rd_kafka_broker_t *rkb,
+                         char *inbuf, size_t inlen,
+                         void **outbuf, size_t *outlenp) {
+
+        rd_kafka_resp_err_t err = RD_KAFKA_RESP_ERR_NO_ERROR;
+        char *decompressed = NULL;
+
+        size_t out_bufsize = ZSTD_getFrameContentSize(inbuf, inlen);
+        switch (out_bufsize) {
+        case ZSTD_CONTENTSIZE_UNKNOWN:
+                /* Decompressed size cannot be determined, make a guess */
+                out_bufsize = inlen * 2;
+        case ZSTD_CONTENTSIZE_ERROR:
+                /* Error calculating frame content size */
+                rd_rkb_dbg(rkb, MSG, "ZSTD",
+                           "Unable to begin ZSTD decompression "
+                           "(out buffer is %zu bytes): %s",
+                           out_bufsize, "Error in determining frame size");
+                return RD_KAFKA_RESP_ERR__BAD_COMPRESSION;
+        default:
+                break;
+        }
+
+        /* Multiple passes to allow resizing output buffer if it's too small */
+        decompressed = rd_malloc(out_bufsize);
+        if (!decompressed) {
+                rd_rkb_dbg(rkb, MSG, "ZSTD",
+                           "Unable to allocate output buffer "
+                           "(%zu bytes): %s",
+                           out_bufsize, rd_strerror(errno));
+                return RD_KAFKA_RESP_ERR__CRIT_SYS_RESOURCE;
+        }
+
+        for (size_t i = 0; i < 3; ++i) {
+                /* Attempt to decompress */
+                size_t ret = ZSTD_decompress(decompressed, out_bufsize, inbuf, inlen);
+                if (!ZSTD_isError(ret)) {
+                        *outlenp = ret;
+                        *outbuf = decompressed;
+                        goto done;
+                }
+
+                /* Check if the destination size is too small */
+                ZSTD_ErrorCode err = ZSTD_getErrorCode(ret);
```

Comment:
variable definititions need to get at the top of the scope.

`err` shadows the top-level `rd_kafka_resp_err_t err`, better call this one `zerr`